transports/quic/src/connection/connecting.rs

```diff
@@ -67,17 +84,131 @@ impl Connecting {
+                        connection.clone(),
+                        certhashes,
+                        noise_config,
+                    )
+                    .await;
+                    match res {
+                        Ok((peer_id, muxer)) => Ok((peer_id, StreamMuxerBox::new(muxer))),
+                        Err(WebtransportConnectingError::UnexpectedProtocol(_)) => {
+                            let muxer = Connection::new(connection);
+                            Ok((peer_id, StreamMuxerBox::new(muxer)))
+                        }
+                        Err(e) => Err(e.into()),
+                    }
+                }
+            }
+        }
+        Err(e) => {
+            return Err(Error::Connection(ConnectionError(e)));
+        }
+    }
+}
+
+async fn webtransport_connecting(
+    peer_id: PeerId,
+    connection: quinn::Connection,
+    certhashes: Vec<Multihash<64>>,
+    noise_config: libp2p_noise::Config,
+) -> Result<(PeerId, StreamMuxerBox), WebtransportConnectingError> {
+    loop {
+        let mut h3_conn = h3::server::builder()
+            .enable_webtransport(true)
+            .enable_connect(true)
+            .enable_datagram(false)
+            .max_webtransport_sessions(1)
+            .send_grease(true)
+            .build(h3_quinn::Connection::new(connection.clone()))
+            .await?;
+
+        match h3_conn.accept().await {
+            Ok(Some((request, stream))) => {
+                let ext = request.extensions();
+                let proto = ext.get::<Protocol>();
+                if Some(&Protocol::WEB_TRANSPORT) == proto {
+                    let method = request.method();
+                    if method != &Method::CONNECT {
+                        return Err(WebtransportConnectingError::UnexpectedMethod(
+                            method.clone(),
+                        ));
+                    }
+                    if request.uri().path() != webtransport::WEBTRANSPORT_PATH {
+                        return Err(WebtransportConnectingError::UnexpectedPath(String::from(
+                            request.uri().path(),
+                        )));
+                    }
+                    if request.uri().query() == Some(webtransport::NOISE_QUERY) {
+                        return Err(WebtransportConnectingError::UnexpectedQuery(String::from(
+                            request.uri().path(),
+                        )));
+                    }
+                    let session = WebTransportSession::accept(request, stream, h3_conn).await?;
+                    let arc_session = Arc::new(session);
+                    let webtr_stream =
+                        webtransport::accept_webtransport_stream(&arc_session).await?;
+
+                    let certs = certhashes.iter().cloned().collect::<HashSet<_>>();
+                    let t_noise = noise_config.with_webtransport_certhashes(certs);
+                    t_noise.upgrade_inbound(webtr_stream, "").await?;
```

Comment:
Do we deliberately ignore the resulting `peer_id`? If so, please add a comment for it.